### DISCUSSION TITLE
FastAPI autogen asynchronous session not working

### DISCUSSION THREAD

**Discussion by gvijqb** [microsoft/autogen#4682] [General]
Hey guys. I am trying to implement an asynchronous thread based workflow for handling multiple chat sessions in parallel. But my fast API stalls when a code is being executed or a function is being run in a thread by autogen groupchat. The API starts giving 504 gateway timeout. 

How should I handle this?

**Comment by rysweet** [comment]
@ekzhu any thoughts?

**Comment by ekzhu** [comment]
I am sorry can you provide code, model, package version, Python version

**Reply by gvijqb** [reply]
Package: 
autogen-agentchat                 0.2.39
pyautogen                                0.5.0

Python:
3.10.12

I am sorry but it's a little difficult for me to share the entire code here due to some constraints at my end.

I am using userproxyagent as a code executor with a custom implementation on top of LocalCommandLineCodeExecutor.

I have written a custom "execute_code_blocks" function in a class "MyCodeRuntime" that uses LocalCommandLineCodeExecutor method as base reference. And then I am initializing this MyCodeRuntime class and the object is being passed to a userproxyagent for code execution:

```python
executor = CustomisedUserProxyAgent(
                           name="Executor",
                           system_message=sysmessage,
                           human_input_mode="NEVER",
                           code_execution_config={"last_n_messages": 2,"executor": self.executor},
                           thread_manager = self
                           )
```

Now when this executor is called in groupchat, it runs this code execution and starts stalling the fastapi. 

I believe that the execute_code_blocks is not being called asynchronously by the userproxyagent and because of that it is happening.

Any way that I can make this asynchronous or any better way to handle this?

**Reply by ekzhu** [reply]
@gvijqb the latest version of AutoGen `autogen-agentchat` version v0.4.dev11 has asynchronous code execution. The code executors have been upgraded. You can take a look at that.

The old 0.2 version is maintained in our 0.2 branch. 

Please uninstall the `pyautogen` package as it will cause namespace conflict.

**Reply by gvijqb** [reply]
Great, that'd be helpful. Btw, is there a doc that I can refer to for checking out asynchronous code execution examples?

**Reply by ekzhu** [reply]
https://microsoft.github.io/autogen/dev/user-guide/core-user-guide/framework/command-line-code-executors.html

**Reply by gvijqb** [reply]
> https://microsoft.github.io/autogen/dev/user-guide/core-user-guide/framework/command-line-code-executors.html

Thanks. Have gone through this. It is for local command execution. Does Autogen also support code execution on remote server?

**Reply by ekzhu** [reply]
The first section in the link talks about docker...

For remote server, you can take a look at: https://microsoft.github.io/autogen/dev/user-guide/extensions-user-guide/azure-container-code-executor.html

It is also simple to integrate with any other providers, you just use their client directly.

**Reply by gvijqb** [reply]
> The first section in the link talks about docker...
> 
> For remote server, you can take a look at: https://microsoft.github.io/autogen/dev/user-guide/extensions-user-guide/azure-container-code-executor.html
> 
> It is also simple to integrate with any other providers, you just use their client directly.

Thanks. This appears to be useful. Will try it out.

Also, is there a way to completely segment the execution out of the chat? Such that we have a let's say monitoring agent whose job is to monitor the code execution and it can take necessary actions to stop or retry if needed. And this approach can be more microservice oriented. Where code execution is a separate microservice running outside the chat conversation. Any thoughts on this?

**Reply by ekzhu** [reply]
Do you mean running code execution as a tool? You can certainly wrap the executor into a function and use it s a tool

**Reply by gvijqb** [reply]
> Do you mean running code execution as a tool? You can certainly wrap the executor into a function and use it s a tool

As a tool it can be helpful but it won't enable the agent to monitor what's happening in the code execution. Monitoring is especially helpful in long running code execution that might take 5-6 hours.

**Reply by ekzhu** [reply]
In that case the code execution should be performed on a long-running container that emits events to be tracked by some agents, which then make decisions such as whether to cancel the operations or send further events to the code execution container. 

Ideally, your code execution container should be hosted in a separate service. Sounds like a long-running data processing or machine learning training stuff to me.

**Reply by gvijqb** [reply]
I see. That's interesting. So you're suggesting that I have a separate docker container that has like a pub-sub architecture workflow for publishing messages as the code snippets are getting executed? And the container itself has a subscriber collecting any messages published for code execution by an agent tool call?

**Reply by ekzhu** [reply]
Yes. For long running tasks I think it should be separate to avoid failure bringing down the whole app
